accord-core/src/main/java/accord/messages/Commit.java

```diff
@@ -275,6 +275,22 @@ public static void commitInvalidate(Node node, TxnId txnId, Unseekables<?> infor
             // TODO (expected, safety): this kind of check needs to be inserted in all equivalent methods
             Invariants.checkState(untilEpoch >= txnId.epoch());
             Invariants.checkState(node.topology().hasEpoch(untilEpoch));
+            // Possible that the invalidation is due to the range no longer existing in untilEpoch, which means that
```

Comment:
So a couple of thoughts here:

1) How can this happen? We should only invoke with an `untilEpoch` we attempted to `Accept` with, and we shouldn't attempt to `Accept` in an epoch where the keys don't exist
2) If we have somehow done this, I think it is fine to contact `untilEpoch` even if some keys are missing? The problem with going to the earliest epoch we can fine that contains them is that if we have somehow gone into a future epoch, we might have left traces we should ideally invalidate for other keys that did still exist, and have moved replica

My preference would be to first figure out (1), as if it's happening it may indicate some other issue that we can address. Then we might anyway want to support a version of `preciseEpochs` that permits topology mismatches.

Reply:
> How can this happen? We should only invoke with an untilEpoch we attempted to Accept with, and we shouldn't attempt to Accept in an epoch where the keys don't exist

This is the stack trace that hits it

```
commitInvalidate:287, Commit$Invalidate (accord.messages)
commitInvalidate:270, Commit$Invalidate (accord.messages)
lambda$commitInvalidate$12:365, Recover (accord.coordinate)
run:-1, 1373949107 (accord.coordinate.Recover$$Lambda$689)
withEpoch:297, Node (accord.local)
commitInvalidate:365, Recover (accord.coordinate)
lambda$invalidate$10:358, Recover (accord.coordinate)
accept:-1, 73688220 (accord.coordinate.Recover$$Lambda$686)
onSuccess:202, Propose$Invalidate (accord.coordinate)
onSuccess:137, Propose$Invalidate (accord.coordinate)
accept:-1, 353023213 (accord.messages.SafeCallback$$Lambda$555)
lambda$safeCall$1:74, SafeCallback (accord.messages)
```

this epoch is coming from `accord.coordinate.Recover#commitInvalidate`

```
Timestamp invalidateUntil = recoverOks.stream().map(ok -> ok.executeAt).reduce(txnId, Timestamp::max);
```

the peers that are sending the `executeAt` have the following for `Command`

```
Command@1850137076{[1,5,2(KW),7]:AcceptedInvalidateWithDefinition}
```

This looks to be triggered by `accord.messages.Accept.Invalidate#apply`


>  I think it is fine to contact untilEpoch even if some keys are missing?

I believe my original logic that Blake approved allowed this, but since you asked for TM to reject these cases, we had to handle here.

Reply:
Ok, so the mistake is that we use a later epoch that is voted on (and recorded as such) on preaccept that wasn't proposed. So I guess we really want to replicate the original epoch we used for sending, which means for any Accept we see, we use the highest timestamp. But for PreAccept we only use TxnId. In this case we *should* never use a future epoch, and we should still invalidate/clear any records we might have created. 

So, something like
`Timestamp invalidateUntil = recoverOks.stream().map(ok -> ok.saveStatus.hasBeen(Accepted) ? ok.executeAt : ok.txnId).reduce(txnId, Timestamp::max);`

This isn't perfect, as an `AcceptInvalidate` could mask an `Accept` with a higher timestamp, and cause us to not clear some state. But I'm not sure it's worth agonising over, as eventually it all gets cleaned up. Better to keep things simple.

Reply:
I pushed this change and see that `CommandsTest` is stable with it, also tried to document the "why" for next authors who review